catalog/core/catalog-core-subscriptionstore/src/main/java/org/codice/ddf/catalog/subscriptionstore/SubscriptionCache.java

```diff
@@ -0,0 +1,234 @@
+        this.factories = new ConcurrentHashMap<>();
+    }
+
+    @Override
+    public void registerSubscriptionFactory(String type, SubscriptionFactory factory) {
+        notEmpty(type, "Subscription type cannot be empty");
+        notNull(factory, "Subscription factory cannot be null");
+
+        if (factories.containsKey(type)) {
+            throw new SubscriptionStoreException(format(
+                    "Factory already exists for subscription type [%s]",
+                    type));
+        }
+
+        factories.put(type, factory);
+    }
+
+    @Nullable
+    @Override
+    public Subscription get(String subscriptionId, String type) {
+        notEmpty(subscriptionId, "subscriptionId cannot be null or empty");
+        notEmpty(type, "type cannot be null or empty");
+
+        if (!subscriptions.containsKey(subscriptionId)) {
+            return null;
+        }
+
+        SubscriptionMetadata metadata = subscriptions.get(subscriptionId);
+
+        if (!type.equals(metadata.getType())) {
+            return null;
+        }
+
+        return metadata.getSubscription()
+                .orElseThrow(() -> new SubscriptionStoreException(format(
+                        "Subscription [%s] was not cached",
+                        metadata.getId())));
+    }
+
+    @Override
+    public String insert(Subscription subscription, String type, String messageBody,
+            String callbackUrl) {
+        notNull(subscription, "subscription cannot be null");
+        notEmpty(type, "type cannot be null or empty");
+        notEmpty(messageBody, "messageBody cannot be null or empty");
+        notEmpty(callbackUrl, "callbackUrl cannot be null or empty");
+
+        SubscriptionMetadata metadata = new SubscriptionMetadata(type, messageBody, callbackUrl);
+
+        createSubscriptionLocally(metadata, subscription);
+        subscriptionStore.insert(metadata);
+
+        return metadata.getId();
+    }
+
+    @Override
+    public void update(Subscription subscription, String type, String messageBody,
+            String callbackUrl, String subscriptionId) {
+        notNull(subscription, "subscription cannot be null");
+        notEmpty(type, "type cannot be null or empty");
+        notEmpty(messageBody, "messageBody cannot be null or empty");
+        notEmpty(callbackUrl, "callbackUrl cannot be null or empty");
+        notEmpty(subscriptionId, "subscriptionId cannot be null or empty");
+
+        if (!contains(subscriptionId, type)) {
+            LOGGER.debug("Target for subscription update [{}|{}] does not exist",
+                    subscriptionId,
+                    type);
+            return;
+        }
+
+        delete(subscriptionId, type);
+
+        SubscriptionMetadata metadata = new SubscriptionMetadata(type,
+                messageBody,
+                callbackUrl,
+                subscriptionId);
+
+        createSubscriptionLocally(metadata, subscription);
+        subscriptionStore.insert(metadata);
```

Comment:
make sure this will fail gracefully maybe move it in front of createSubscriptionLocally if that doesn't have much logic that would cause a failure